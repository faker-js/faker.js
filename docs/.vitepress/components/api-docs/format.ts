export function formatResult(result: unknown): string {
  return result === undefined
    ? 'undefined'
    : typeof result === 'bigint'
      ? `${result}n`
      : JSON.stringify(result, undefined, 2)
          .replaceAll('\\r', '')
          .replaceAll('<', '&lt;')
          .replaceAll(
            /(^ *|: )"([^'\n]*?)"(?=,?$|: )/gm,
            (_, p1, p2) => `${p1}'${p2.replace(/\\"/g, '"')}'`
          )
          .replaceAll(/\n */g, ' ');
}
