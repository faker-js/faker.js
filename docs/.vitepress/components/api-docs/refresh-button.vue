<script setup lang="ts">
import { ref } from 'vue';

// This should probably use emit instead, but emit cannot be awaited
const { refresh } = defineProps<{ refresh: () => Promise<void> }>();

const spinning = ref(false);

async function onRefresh() {
  spinning.value = true;
  await Promise.all([refresh(), delay(100)]);
  spinning.value = false;
}

// Extra delay to make the spinning effect more visible
// Some examples barely/don't change, so the spinning is the only visible effect
function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
</script>

<template>
  <button
    class="refresh"
    title="Refresh Examples"
    :disabled="spinning"
    @click="onRefresh"
  >
    <div :class="{ spinning: spinning }" />
  </button>
</template>

<style scoped>
button.refresh {
  border: 1px solid var(--vp-code-copy-code-border-color);
  border-radius: 4px;
  width: 40px;
  height: 40px;
  font-size: 25px;
  vertical-align: middle;
}

button.refresh div {
  background-image: url('refresh.svg');
  background-position: 50%;
  background-size: 20px;
  background-repeat: no-repeat;
  width: 100%;
  height: 100%;
}

button.refresh:hover {
  background-color: var(--vp-code-copy-code-bg);
  opacity: 1;
}

div.spinning {
  animation: spin 1s linear infinite;
}

@keyframes spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
</style>
