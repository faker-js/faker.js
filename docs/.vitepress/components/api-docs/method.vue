<script setup lang="ts">
import { computed, ref, useTemplateRef } from 'vue';
import { sourceBaseUrl } from '../../../api/source-base-url';
import { slugify } from '../../shared/utils/slugify';
import { formatResult } from './format';
import type { ApiDocsMethod } from './method';
import MethodParameters from './method-parameters.vue';
import RefreshButton from './refresh-button.vue';

const { method } = defineProps<{ method: ApiDocsMethod }>();
const {
  deprecated,
  description,
  since,
  parameters,
  returns,
  throws,
  signature,
  examples,
  refresh,
  seeAlsos,
  sourcePath,
} = method;

const code = useTemplateRef('code');
const codeBlock = computed(() => code.value?.querySelector('div pre code'));
const codeLines = ref<Element[]>();

function initRefresh(): Element[] {
  if (codeBlock.value == null) {
    return [];
  }
  const domLines = codeBlock.value.querySelectorAll('.line');
  let lineIndex = 0;
  const result: Element[] = [];
  while (lineIndex < domLines.length) {
    // Skip empty and preparatory lines (no '^faker.' invocation)
    if (
      domLines[lineIndex]?.children.length === 0 ||
      !/^\w*faker\w*\./i.test(domLines[lineIndex]?.textContent ?? '')
    ) {
      lineIndex++;
      continue;
    }

    // Skip to end of the invocation (if multiline)
    while (
      domLines[lineIndex] != null &&
      !/^([^ ].*)?\)(\.\w+)?;? ?(\/\/|$)/.test(
        domLines[lineIndex]?.textContent ?? ''
      )
    ) {
      lineIndex++;
    }

    if (lineIndex >= domLines.length) {
      break;
    }

    const domLine = domLines[lineIndex];
    result.push(domLine);
    lineIndex++;

    // Purge old results
    if (domLine.lastElementChild?.textContent?.startsWith('//')) {
      // Inline comments
      domLine.lastElementChild.remove();
    } else {
      // Multiline comments
      while (domLines[lineIndex]?.children[0]?.textContent?.startsWith('//')) {
        domLines[lineIndex].previousSibling?.remove(); // newline
        domLines[lineIndex].remove(); // comment
        lineIndex++;
      }
    }

    // Add space between invocation and comment (if missing)
    const lastElementChild = domLine.lastElementChild;
    if (
      lastElementChild != null &&
      !lastElementChild.textContent?.endsWith(' ')
    ) {
      lastElementChild.textContent += ' ';
    }
  }

  return result;
}

async function onRefresh(): Promise<void> {
  if (refresh != null && codeBlock.value != null) {
    codeLines.value ??= initRefresh();

    const results = await refresh();

    // Remove old comments
    codeBlock.value
      .querySelectorAll('.comment-delete-marker')
      .forEach((el) => el.remove());

    // Insert new comments
    for (let i = 0; i < results.length; i++) {
      const result = results[i];
      const domLine = codeLines.value[i];
      const prettyResult = formatResult(result);
      const resultLines = prettyResult.split('\\n');

      if (resultLines.length === 1) {
        domLine.insertAdjacentHTML('beforeend', newCommentSpan(resultLines[0]));
      } else {
        for (const line of resultLines.reverse()) {
          domLine.insertAdjacentHTML('afterend', newCommentLine(line));
        }
      }
    }
  }
}

function newCommentLine(content: string): string {
  return `<span class="line comment-delete-marker">\n${newCommentSpan(content)}</span>`;
}

function newCommentSpan(content: string): string {
  return `<span class="comment-delete-marker" style="--shiki-light:#6A737D;--shiki-dark:#6A737D">// ${content}</span>`;
}

function seeAlsoToUrl(see: string): string {
  const [, module, methodName] = see.replace(/\(.*/, '').split('\.');

  if (!methodName) {
    return `faker.html#${slugify(module)}`;
  }

  return `${module}.html#${slugify(methodName)}`;
}
</script>

<template>
  <div>
    <div v-if="deprecated" class="warning custom-block">
      <p class="custom-block-title">Deprecated</p>
      <p>This method is deprecated and will be removed in a future version.</p>
      <span v-html="deprecated" />
    </div>

    <div v-html="description"></div>

    <p v-if="since">
      <em>Available since v{{ since }}</em>
    </p>

    <MethodParameters v-if="parameters.length > 0" :parameters="parameters" />

    <p><strong>Returns:</strong> {{ returns }}</p>

    <p v-if="throws"><strong>Throws:</strong> <span v-html="throws" /></p>

    <div v-html="signature" />

    <h3 class="inline">Examples</h3>
    <RefreshButton
      class="refresh"
      v-if="refresh != null"
      style="margin-left: 0.5em"
      :refresh="onRefresh"
    />
    <div ref="code" v-html="examples" />

    <div v-if="seeAlsos.length > 0">
      <h3>See Also</h3>
      <ul>
        <li v-for="seeAlso of seeAlsos" :key="seeAlso">
          <a
            v-if="seeAlso.startsWith('faker.')"
            :href="seeAlsoToUrl(seeAlso)"
            v-html="seeAlso"
          />
          <div v-else v-html="seeAlso" />
        </li>
      </ul>
    </div>

    <div v-if="sourcePath">
      <h3>Source</h3>
      <ul>
        <li>
          <a
            :href="sourceBaseUrl + sourcePath"
            target="_blank"
            class="source-link"
          >
            View Source
            <svg
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 24 24"
              fill="currentColor"
              width="1.2em"
              height="1.2em"
              class="source-link-icon"
            >
              <path
                d="M14,3V5H17.59L7.76,14.83L9.17,16.24L19,6.41V10H21V3M19,19H5V5H12V3H5C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V12H19V19Z"
              />
            </svg>
          </a>
        </li>
      </ul>
    </div>
  </div>
</template>

<style scoped>
a.source-link {
  display: flex;
  align-items: center;
}

svg.source-link-icon {
  display: inline;
  margin-left: 0.3em;
}

h3.inline {
  display: inline-block;
}
</style>
