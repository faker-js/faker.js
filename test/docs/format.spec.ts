import { describe, expect, it } from 'vitest';
import { formatResult } from '../../docs/.vitepress/components/api-docs/format';

describe('formatResult', () => {
  it('should format undefined', () => {
    const value = undefined;
    const actual = formatResult(value);

    expect(actual).toBeTypeOf('string');
    expect(actual).toBe('undefined');
    expect(actual).toMatchSnapshot();
  });

  it('should format bigint', () => {
    const actual = formatResult(135464154865415n);

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format object', () => {
    const actual = formatResult({ a: 1, b: '2' });

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format array', () => {
    const actual = formatResult([1, '2']);

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format string', () => {
    const actual = formatResult('a simple string');

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format string with special characters', () => {
    const actual = formatResult('string with "special" characters');

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format string with new lines', () => {
    const actual = formatResult('string\nwith\nnew\nlines');

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format number', () => {
    const actual = formatResult(123);

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });

  it('should format Date', () => {
    const actual = formatResult(new Date(Date.UTC(2025, 0, 1)));

    expect(actual).toBeTypeOf('string');
    expect(actual).toMatchSnapshot();
  });
});
