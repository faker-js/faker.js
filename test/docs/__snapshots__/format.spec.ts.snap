// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatResult > should format Date 1`] = `"'2025-01-01T00:00:00.000Z'"`;

exports[`formatResult > should format array 1`] = `"[ 1, '2' ]"`;

exports[`formatResult > should format bigint 1`] = `"135464154865415n"`;

exports[`formatResult > should format number 1`] = `"123"`;

exports[`formatResult > should format object 1`] = `"{ 'a': 1, 'b': '2' }"`;

exports[`formatResult > should format string 1`] = `"'a simple string'"`;

exports[`formatResult > should format string with new lines 1`] = `"'string\\nwith\\nnew\\nlines'"`;

exports[`formatResult > should format string with special characters 1`] = `"'string with "special" characters'"`;

exports[`formatResult > should format undefined 1`] = `"undefined"`;
